README.md
[![PyPI version](https://badge.fury.io/py/sentence-embedding-evaluation-german.svg)](https://badge.fury.io/py/sentence-embedding-evaluation-german)
[![Total alerts](https://img.shields.io/lgtm/alerts/g/linguistik/sentence-embedding-evaluation-german.svg?logo=lgtm&logoWidth=18)](https://lgtm.com/projects/g/linguistik/sentence-embedding-evaluation-german/alerts/)
[![Language grade: Python](https://img.shields.io/lgtm/grade/python/g/linguistik/sentence-embedding-evaluation-german.svg?logo=lgtm&logoWidth=18)](https://lgtm.com/projects/g/linguistik/sentence-embedding-evaluation-german/context:python)

# sentence-embedding-evaluation-german
Sentence embedding evaluation for German.

This library is inspired by [SentEval](https://github.com/facebookresearch/SentEval) but focuses on German language downstream tasks.


## Downstream tasks

| task | type | properties | \#train | \#test | target | info |
|:---:|:---:|:---:|:---:|:---:|:---:|:---:|
| TOXIC | 👿 toxic comments | facebook comments | 2596 | 944 | binary {0,1} | GermEval 2021, comments subtask 1, [📁](https://github.com/germeval2021toxic/SharedTask) [📖](https://aclanthology.org/2021.germeval-1.1) |
| ENGAGE | 🤗 engaging comments | facebook comments | 2596 | 944 | binary {0,1} | GermEval 2021, comments subtask 2, [📁](https://github.com/germeval2021toxic/SharedTask) [📖](https://aclanthology.org/2021.germeval-1.1) |
| FCLAIM | ☝️ fact-claiming comments | facebook comments | 2596 | 944 | binary {0,1} | GermEval 2021, comments subtask 3, [📁](https://github.com/germeval2021toxic/SharedTask) [📖](https://aclanthology.org/2021.germeval-1.1) |
| VMWE | verbal idioms | newspaper | 6651 | 1446 | binary (figuratively, literally) | GermEval 2021, verbal idioms, [📁](https://github.com/rafehr/vid-disambiguation-sharedtask) [📖](https://aclanthology.org/2020.figlang-1.29.pdf) |
| OL19A | 👿 offensive language | tweets | 3184 | 3030 | binary {0,1} | GermEval 2018, [📁](https://projects.fzai.h-da.de/iggsa/data-2019/) |
| OL19B | 👿 offensive language, fine-grained | tweets | 3184 | 3030 | 4 catg. (profanity, insult, abuse, oth.) | GermEval 2018, [📁](https://projects.fzai.h-da.de/iggsa/data-2019/) |
| OL19C | 👿 explicit vs. implicit offense | tweets | 1920 | 929 | binary (explicit, implicit) | GermEval 2018, [📁](https://projects.fzai.h-da.de/iggsa/data-2019/) |
| OL18A | 👿 offensive language | tweets | 4007 | 3397 | binary {0,1} | GermEval 2018, [📁](https://github.com/uds-lsv/GermEval-2018-Data) |
| OL18B | 👿 offensive language, fine-grained | tweets | 4007 | 3397 | 4 catg. (profanity, insult, abuse, oth.) | GermEval 2018, [📁](https://github.com/uds-lsv/GermEval-2018-Data) |
| ABSD1 | 🛤️ relevance classification | Deutsche Bahn customer feedback | 19431 | 2565 | binary | GermEval 2017, [📁](https://sites.google.com/view/germeval2017-absa/data) |
| ABSD2 | 🛤️ sentiment analysis | Deutsche Bahn customer feedback | 19431 | 2565 | 3 catg. (pos., neg., neutral) | GermEval 2017, [📁](https://sites.google.com/view/germeval2017-absa/data) |
| ABSD3 | 🛤️ aspect categories | Deutsche Bahn customer feedback | 19431 | 2565 | 20 catg. | GermEval 2017, [📁](https://sites.google.com/view/germeval2017-absa/data) |


## Download datasets

```sh
bash download-datasets.sh
```

## Usage example

```py
from typing import List
import sentence_embedding_evaluation_german as seeg
import torch

# (1) Instantiate your Embedding model
emb_dim = 512
vocab_sz = 128
emb = torch.randn((vocab_sz, emb_dim), requires_grad=False)
emb = torch.nn.Embedding.from_pretrained(emb)
assert emb.weight.requires_grad == False

# (2) Specify the preprocessing
def preprocesser(batch: List[str], params: dict=None) -> List[List[float]]:
    """ Specify your embedding or pretrained encoder here
    Paramters:
    ----------
    params : dict
        The params dictionary
    batch : List[str]
        A list of sentence as string
    Returns:
    --------
    List[List[float]]
        A list of embedding vectors
    """
    features = []
    for sent in batch:
        try:
            ids = torch.tensor([ord(c) % 128 for c in sent])
        except:
            print(sent)
        h = emb(ids)
        features.append(h.mean(axis=0))
    features = torch.stack(features, dim=0)
    return features

# (3) Training settings
params = {
    'datafolder': '../datasets',
    'batch_size': 128, 
    'num_epochs': 20,
    # 'early_stopping': True,
    # 'split_ratio': 0.2,  # if early_stopping=True
    # 'patience': 5,  # if early_stopping=True
}

# (4) Specify downstream tasks
downstream_tasks = [
    'TOXIC', 'ENGAGE', 'FCLAIM', 'VMWE',
    'OL19A', 'OL19B', 'OL19C',
    'ABSD1', 'ABSD2', 'ABSD3']

# (5) Run experiments
results = seeg.evaluate(downstream_tasks, preprocesser, **params)
```

## Appendix

### Installation
The `sentence-embedding-evaluation-german` [git repo](http://github.com/linguistik/sentence-embedding-evaluation-german) is available as [PyPi package](https://pypi.org/project/sentence-embedding-evaluation-german)

```sh
pip install sentence-embedding-evaluation-german
pip install git+ssh://git@github.com/linguistik/sentence-embedding-evaluation-german.git
```

### Install a virtual environment

```sh
python3.7 -m venv .venv
source .venv/bin/activate
pip install --upgrade pip
pip install -r requirements.txt --no-cache-dir
pip install -r requirements-dev.txt --no-cache-dir
pip install -r requirements-demo.txt --no-cache-dir
```

(If your git repo is stored in a folder with whitespaces, then don't use the subfolder `.venv`. Use an absolute path without whitespaces.)

### Python commands

* Jupyter for the examples: `jupyter lab`
* Check syntax: `flake8 --ignore=F401 --exclude=$(grep -v '^#' .gitignore | xargs | sed -e 's/ /,/g')`

Publish

```sh
pandoc README.md --from markdown --to rst -s -o README.rst
python setup.py sdist 
twine upload -r pypi dist/*
```

### Clean up 

```sh
find . -type f -name "*.pyc" | xargs rm
find . -type d -name "__pycache__" | xargs rm -r
rm -r .pytest_cache
rm -r .venv
```


### Support
Please [open an issue](https://github.com/linguistik/sentence-embedding-evaluation-german/issues/new) for support.


### Contributing
Please contribute using [Github Flow](https://guides.github.com/introduction/flow/). Create a branch, add commits, and [open a pull request](https://github.com/linguistik/sentence-embedding-evaluation-german/compare/).


CHANGES.md

# 0.1.0 / 2022-04-08

  * Initial Release


download-datasets.sh
#!/bin/bash

mkdir -p datasets/germeval21
wget -q "https://raw.githubusercontent.com/germeval2021toxic/SharedTask/main/Data%20Sets/GermEval21_TrainData.csv" -O datasets/germeval21/train.csv
wget -q "https://raw.githubusercontent.com/germeval2021toxic/SharedTask/main/Data%20Sets/GermEval21_TestData.csv" -O datasets/germeval21/test.csv

mkdir -p datasets/germeval21vmwe
wget -q "https://raw.githubusercontent.com/rafehr/vid-disambiguation-sharedtask/main/data/train/train.tsv" -O datasets/germeval21vmwe/train.tsv
wget -q "https://raw.githubusercontent.com/rafehr/vid-disambiguation-sharedtask/main/data/test/test.tsv" -O datasets/germeval21vmwe/test.tsv

mkdir -p datasets/germeval19
wget -q "https://projects.fzai.h-da.de/iggsa/wp-content/uploads/2019/09/germeval2019.training_subtask1_2_korrigiert.txt" -O datasets/germeval19/train12.txt
wget -q "https://projects.fzai.h-da.de/iggsa/wp-content/uploads/2019/05/germeval2019.training_subtask3.txt" -O datasets/germeval19/train3.txt
wget -q "https://projects.fzai.h-da.de/iggsa/wp-content/uploads/2019/08/germeval2019GoldLabelsSubtask1_2.txt" -O datasets/germeval19/gold12.txt
wget -q "https://projects.fzai.h-da.de/iggsa/wp-content/uploads/2019/08/germeval2019GoldLabelsSubtask3.txt" -O datasets/germeval19/gold3.txt

mkdir -p datasets/germeval18
wget -q "https://projects.fzai.h-da.de/iggsa/wp-content/uploads/2019/05/germeval2018.training.txt" -O datasets/germeval18/train.txt
wget -q "https://projects.fzai.h-da.de/iggsa/wp-content/uploads/2019/05/germeval2018.test_.txt" -O datasets/germeval18/test.txt

mkdir -p datasets/germeval17
wget -q "http://ltdata1.informatik.uni-hamburg.de/germeval2017/train-2017-09-15.tsv" -O datasets/germeval17/train.tsv
wget -q "http://ltdata1.informatik.uni-hamburg.de/germeval2017/test_syn-2017-09-15.tsv" -O datasets/germeval17/test.tsv
# wget -q "http://ltdata1.informatik.uni-hamburg.de/germeval2017/test_dia-2017-09-15.tsv" -O datasets/germeval17/test-dia.tsv



sentence_embedding_evaluation_german/data.py
import pandas as pd
import numpy as np
import torch


def get_data_split(n: int,
                   split_ratio: float = 0.2,
                   random_seed: int = 42):
    # set seed for reproducibility
    if random_seed:
        torch.manual_seed(random_seed)
    # random indicies
    idx = torch.randperm(n)
    n_valid = int(n * split_ratio)
    idx_valid = idx[:n_valid]
    idx_train = idx[n_valid:]
    return idx_train, idx_valid


class GermEval17(torch.utils.data.Dataset):
    """ ABSD-Relevance, -Sentiment, -Category
    Examples:
    ---------
    dset = GermEval17(
        preprocesser, "Relevance", test=False,
        early_stopping=True, split_ratio=0.1)
    X_valid, y_valid = dset.get_validation_set()
    n_classes = dset.num_classes()
    dgen = torch.utils.data.DataLoader(
        dset, **{'batch_size': 64, 'shuffle': True, 'num_workers': 6})
    for X, y in dgen: break
    """
    def __init__(self,
                 preprocesser,
                 datafolder: str = "datasets",
                 task: int = 1,
                 test: bool = False,
                 early_stopping: bool = False,
                 split_ratio: float = 0.2,
                 random_seed: int = 42):
        assert task in ["Relevance", "Sentiment", "Category"]
        self.colidx = int(
            ["Relevance", "Sentiment", "Category"].index(task) + 2)

        if task == "Relevance":
            self.labels = [False, True]
        elif task == "Sentiment":
            self.labels = ['negative', 'neutral', 'positive']
        elif task == "Category":
            self.labels = [
                'Image', 'Informationen', 'Connectivity',
                'Auslastung_und_Platzangebot', 'Service_und_Kundenbetreuung',
                'Gastronomisches_Angebot', 'Allgemein', 'Design',
                'Sonstige_Unregelmässigkeiten', 'Gepäck', 'DB_App_und_Website',
                'Atmosphäre', 'Zugfahrt', 'Ticketkauf', 'nan',
                'Reisen_mit_Kindern', 'Barrierefreiheit', 'Sicherheit',
                'Toiletten', 'Komfort_und_Ausstattung']

        # read data
        split = "test" if test else "train"
        data = pd.read_csv(
            f"{datafolder}/germeval17/{split}.tsv", sep="\t").values
        data[:, 4] = [str(s).split(":")[0] for s in data[:, 4]]
        # bad examples to be removed
        idxbad = [i for i, x in enumerate(data[:, 1])
                  if not isinstance(x, str)]
        data = np.delete(data, idxbad, axis=0)

        # preprocess
        self.X = preprocesser(data[:, 1].tolist())
        self.y = torch.tensor(
            [self.labels.index(row[self.colidx]) for row in data])
        # prepare data split
        if early_stopping and split == "train":
            self.indices, self.idx_valid = get_data_split(
                self.X.shape[0], random_seed=random_seed)
        else:
            self.indices = torch.tensor(range(self.X.shape[0]))
            self.idx_valid = None

    def get_validation_set(self):
        if self.idx_valid is not None:
            return self.X[self.idx_valid], self.y[self.idx_valid]
        else:
            return None, None

    def num_classes(self):
        return len(self.labels)

    def num_features(self):
        return self.X.shape[1]

    def __len__(self):
        return len(self.indices)

    def __getitem__(self, rowidx):
        return self.X[self.indices[rowidx]], self.y[self.indices[rowidx]]


class GermEval18(torch.utils.data.Dataset):
    """ OL18A, OL18B
    Examples:
    ---------
    dset = GermEval18(
        preprocesser, "A", test=False,
        early_stopping=True, split_ratio=0.1)
    X_valid, y_valid = dset.get_validation_set()
    n_classes = dset.num_classes()
    dgen = torch.utils.data.DataLoader(
        dset, **{'batch_size': 64, 'shuffle': True, 'num_workers': 6})
    for X, y in dgen: break
    """
    def __init__(self,
                 preprocesser,
                 datafolder: str = "datasets",
                 task: int = 1,
                 test: bool = False,
                 early_stopping: bool = False,
                 split_ratio: float = 0.2,
                 random_seed: int = 42):
        assert task in ["A", "B", "C"]
        self.colidx = int(["A", "B", "C"].index(task) + 1)

        if task == "A":
            self.labels = ['OFFENSE', 'OTHER']
        elif task == "B":
            self.labels = ['PROFANITY', 'INSULT', 'ABUSE', 'OTHER']

        # read data
        split = "test" if test else "train"
        data = pd.read_csv(
            f"{datafolder}/germeval18/{split}.txt", sep="\t").values
        # preprocess
        self.X = preprocesser(data[:, 0].tolist())
        self.y = torch.tensor(
            [self.labels.index(row[self.colidx]) for row in data])

        # prepare data split
        if early_stopping and split == "train":
            self.indices, self.idx_valid = get_data_split(
                self.X.shape[0], random_seed=random_seed)
        else:
            self.indices = torch.tensor(range(self.X.shape[0]))
            self.idx_valid = None

    def get_validation_set(self):
        if self.idx_valid is not None:
            return self.X[self.idx_valid], self.y[self.idx_valid]
        else:
            return None, None

    def num_classes(self):
        return len(self.labels)

    def num_features(self):
        return self.X.shape[1]

    def __len__(self):
        return len(self.indices)

    def __getitem__(self, rowidx):
        return self.X[self.indices[rowidx]], self.y[self.indices[rowidx]]


class GermEval19(torch.utils.data.Dataset):
    """ OL19A, OL19B, OL19C
    Examples:
    ---------
    dset = GermEval19(
        preprocesser, "A", test=False,
        early_stopping=True, split_ratio=0.1)
    X_valid, y_valid = dset.get_validation_set()
    n_classes = dset.num_classes()
    dgen = torch.utils.data.DataLoader(
        dset, **{'batch_size': 64, 'shuffle': True, 'num_workers': 6})
    for X, y in dgen: break
    """
    def __init__(self,
                 preprocesser,
                 datafolder: str = "datasets",
                 task: int = 1,
                 test: bool = False,
                 early_stopping: bool = False,
                 split_ratio: float = 0.2,
                 random_seed: int = 42):
        assert task in ["A", "B", "C"]
        self.colidx = int(["A", "B", "C"].index(task) + 1)

        if task == "A":
            self.labels = ['OFFENSE', 'OTHER']
            fsuf = "12"
        elif task == "B":
            self.labels = ['PROFANITY', 'INSULT', 'ABUSE', 'OTHER']
            fsuf = "12"
        elif task == "C":
            self.labels = ['EXPLICIT', 'IMPLICIT']
            fsuf = "3"

        # read data
        split = "gold" if test else "train"
        data = pd.read_csv(
            f"{datafolder}/germeval19/{split}{fsuf}.txt", sep="\t").values
        # preprocess
        self.X = preprocesser(data[:, 0].tolist())
        self.y = torch.tensor(
            [self.labels.index(row[self.colidx]) for row in data])

        # prepare data split
        if early_stopping and split == "train":
            self.indices, self.idx_valid = get_data_split(
                self.X.shape[0], random_seed=random_seed)
        else:
            self.indices = torch.tensor(range(self.X.shape[0]))
            self.idx_valid = None

    def get_validation_set(self):
        if self.idx_valid is not None:
            return self.X[self.idx_valid], self.y[self.idx_valid]
        else:
            return None, None

    def num_classes(self):
        return len(self.labels)

    def num_features(self):
        return self.X.shape[1]

    def __len__(self):
        return len(self.indices)

    def __getitem__(self, rowidx):
        return self.X[self.indices[rowidx]], self.y[self.indices[rowidx]]


class GermEval21(torch.utils.data.Dataset):
    """ TOXIC, ENGAGE FCLAIM
    Examples:
    ---------
    dset = GermEval21(
        preprocesser, "TOXIC", test=False,
        early_stopping=True, split_ratio=0.1)
    X_valid, y_valid = dset.get_validation_set()
    n_classes = dset.num_classes()
    dgen = torch.utils.data.DataLoader(
        dset, **{'batch_size': 64, 'shuffle': True, 'num_workers': 6})
    for X, y in dgen: break
    """
    def __init__(self,
                 preprocesser,
                 datafolder: str = "datasets",
                 task: str = "TOXIC",
                 test: bool = False,
                 early_stopping: bool = False,
                 split_ratio: float = 0.2,
                 random_seed: int = 42):
        assert task in ["TOXIC", "ENGAGE", "FCLAIM"]
        self.colidx = int(["TOXIC", "ENGAGE", "FCLAIM"].index(task) + 2)

        # read data
        split = "test" if test else "train"
        data = pd.read_csv(f"{datafolder}/germeval21/{split}.csv").values
        # preprocess
        self.X = preprocesser(data[:, 1].tolist())
        self.y = torch.tensor([row[self.colidx] for row in data])

        # prepare data split
        if early_stopping and split == "train":
            self.indices, self.idx_valid = get_data_split(
                self.X.shape[0], random_seed=random_seed)
        else:
            self.indices = torch.tensor(range(self.X.shape[0]))
            self.idx_valid = None

    def get_validation_set(self):
        if self.idx_valid is not None:
            return self.X[self.idx_valid], self.y[self.idx_valid]
        else:
            return None, None

    def num_classes(self):
        return 2

    def num_features(self):
        return self.X.shape[1]

    def __len__(self):
        return len(self.indices)

    def __getitem__(self, rowidx):
        return self.X[self.indices[rowidx]], self.y[self.indices[rowidx]]


class GermEval21vmwe(torch.utils.data.Dataset):
    """ VMWE
    Examples:
    ---------
    dset = GermEval21vmwe(
        preprocesser, test=False,
        early_stopping=True, split_ratio=0.1)
    X_valid, y_valid = dset.get_validation_set()
    n_classes = dset.num_classes()
    dgen = torch.utils.data.DataLoader(
        dset, **{'batch_size': 64, 'shuffle': True, 'num_workers': 6})
    for X, y in dgen: break
    """
    def __init__(self,
                 preprocesser,
                 datafolder: str = "datasets",
                 test: bool = False,
                 early_stopping: bool = False,
                 split_ratio: float = 0.2,
                 random_seed: int = 42):
        # self.labels = ['figuratively', 'literally', 'both', 'undecidable']
        self.labels = ['figuratively', 'literally']

        # read data
        split = "test" if test else "train"
        data = pd.read_csv(
            f"{datafolder}/germeval21vmwe/{split}.tsv", sep="\t").values
        # bad examples to be removed
        idxbad = [i for i, x in enumerate(data[:, 2])
                  if x not in self.labels]
        data = np.delete(data, idxbad, axis=0)

        # preprocess
        self.X = preprocesser(data[:, 3].tolist())
        self.y = torch.tensor([self.labels.index(row[2]) for row in data])

        # prepare data split
        if early_stopping and split == "train":
            self.indices, self.idx_valid = get_data_split(
                self.X.shape[0], random_seed=random_seed)
        else:
            self.indices = torch.tensor(range(self.X.shape[0]))
            self.idx_valid = None

    def get_validation_set(self):
        if self.idx_valid is not None:
            return self.X[self.idx_valid], self.y[self.idx_valid]
        else:
            return None, None

    def num_classes(self):
        return 2

    def num_features(self):
        return self.X.shape[1]

    def __len__(self):
        return len(self.indices)

    def __getitem__(self, rowidx):
        return self.X[self.indices[rowidx]], self.y[self.indices[rowidx]]


sentence_embedding_evaluation_german/model.py
import torch
import types
from typing import List
import sklearn.metrics
from .data import (
    GermEval17, GermEval18, GermEval19, GermEval21, GermEval21vmwe)


class ClassiferModel(torch.nn.Module):
    def __init__(self,
                 input_size: int,
                 output_size: int,  # nclasses
                 bias: bool = True,
                 *args, **kwargs):
        super(ClassiferModel, self).__init__(*args, **kwargs)
        self.bias = bias
        self.final = torch.nn.Linear(
            input_size, output_size, bias=bias)
        self.soft = torch.nn.Softmax(dim=1)
        self._reset_parameters()

    def _reset_parameters(self) -> None:
        torch.manual_seed(42)
        torch.nn.init.xavier_normal_(self.final.weight, gain=1.0)
        if self.bias:
            torch.nn.init.zeros_(self.final.bias)

    def forward(self, inputs: torch.Tensor) -> torch.Tensor:
        x = self.final(inputs)
        return self.soft(x)


def build_model(**kwargs):
    return ClassiferModel(
        input_size=kwargs['n_features'],
        output_size=kwargs['n_classes'])


def evaluate(downstream_tasks: List[str],
             preprocesser: types.FunctionType,
             modelbuilder: types.FunctionType = None,
             datafolder: str = "./datasets",
             batch_size: int = 64,
             num_epochs: int = 20,
             early_stopping: bool = False,
             split_ratio: float = 0.2,
             patience: int = 5):
    # start
    results = []
    for downstream_task in downstream_tasks:
        # load datasets
        if downstream_task == "ABSD1":
            ds_train = GermEval17(
                preprocesser, datafolder=datafolder,
                task="Relevance", test=False, split_ratio=split_ratio,
                early_stopping=early_stopping)
            ds_test = GermEval17(
                preprocesser, datafolder=datafolder,
                task="Relevance", test=True)
        elif downstream_task == "ABSD2":
            ds_train = GermEval17(
                preprocesser, datafolder=datafolder,
                task="Sentiment", test=False, split_ratio=split_ratio,
                early_stopping=early_stopping)
            ds_test = GermEval17(
                preprocesser, datafolder=datafolder,
                task="Sentiment", test=True)
        elif downstream_task == "ABSD3":
            ds_train = GermEval17(
                preprocesser, datafolder=datafolder,
                task="Category", test=False, split_ratio=split_ratio,
                early_stopping=early_stopping)
            ds_test = GermEval17(
                preprocesser, datafolder=datafolder,
                task="Category", test=True)

        elif downstream_task == "OL18A":
            ds_train = GermEval18(
                preprocesser, datafolder=datafolder,
                task="A", test=False, split_ratio=split_ratio,
                early_stopping=early_stopping)
            ds_test = GermEval18(
                preprocesser, datafolder=datafolder,
                task="A", test=True)
        elif downstream_task == "OL18B":
            ds_train = GermEval18(
                preprocesser, datafolder=datafolder,
                task="B", test=False, split_ratio=split_ratio,
                early_stopping=early_stopping)
            ds_test = GermEval18(
                preprocesser, datafolder=datafolder,
                task="B", test=True)

        elif downstream_task == "OL19A":
            ds_train = GermEval19(
                preprocesser, datafolder=datafolder,
                task="A", test=False, split_ratio=split_ratio,
                early_stopping=early_stopping)
            ds_test = GermEval19(
                preprocesser, datafolder=datafolder,
                task="A", test=True)
        elif downstream_task == "OL19B":
            ds_train = GermEval19(
                preprocesser, datafolder=datafolder,
                task="B", test=False, split_ratio=split_ratio,
                early_stopping=early_stopping)
            ds_test = GermEval19(
                preprocesser, datafolder=datafolder,
                task="B", test=True)
        elif downstream_task == "OL19C":
            ds_train = GermEval19(
                preprocesser, datafolder=datafolder,
                task="C", test=False, split_ratio=split_ratio,
                early_stopping=early_stopping)
            ds_test = GermEval19(
                preprocesser, datafolder=datafolder,
                task="C", test=True)

        elif downstream_task == "TOXIC":
            ds_train = GermEval21(
                preprocesser, datafolder=datafolder,
                task="TOXIC", test=False, split_ratio=split_ratio,
                early_stopping=early_stopping)
            ds_test = GermEval21(
                preprocesser, datafolder=datafolder,
                task="TOXIC", test=True)
        elif downstream_task == "ENGAGE":
            ds_train = GermEval21(
                preprocesser, datafolder=datafolder,
                task="ENGAGE", test=False, split_ratio=split_ratio,
                early_stopping=early_stopping)
            ds_test = GermEval21(
                preprocesser, datafolder=datafolder,
                task="ENGAGE", test=True)
        elif downstream_task == "FCLAIM":
            ds_train = GermEval21(
                preprocesser, datafolder=datafolder,
                task="FCLAIM", test=False, split_ratio=split_ratio,
                early_stopping=early_stopping)
            ds_test = GermEval21(
                preprocesser, datafolder=datafolder,
                task="FCLAIM", test=True)

        elif downstream_task == "VMWE":
            ds_train = GermEval21vmwe(
                preprocesser, datafolder=datafolder,
                test=False, split_ratio=split_ratio,
                early_stopping=early_stopping)
            ds_test = GermEval21vmwe(
                preprocesser, datafolder=datafolder,
                test=True)
        else:
            raise Exception(
                f"Downstream task '{downstream_task}' not available.")

        # data loader
        X_valid, y_valid = ds_train.get_validation_set()
        n_classes = ds_train.num_classes()
        n_features = ds_train.num_features()
        dgen = torch.utils.data.DataLoader(
            ds_train, batch_size=batch_size, shuffle=False)

        # init new model
        if not isinstance(modelbuilder, types.FunctionType):
            modelbuilder = build_model

        model = modelbuilder(
            n_features=n_features,
            n_classes=n_classes)

        # loss function: balanced
        loss_fn = torch.nn.CrossEntropyLoss(
            weight=torch.ones(n_classes) / n_classes,
            reduction='mean'
        )

        # early stopping
        if y_valid is not None:
            with torch.no_grad():
                valid_loss = loss_fn(model(X_valid), y_valid).item()
                wait = 0

        # optimization settings
        optimizer = torch.optim.Adam(
            model.parameters(), lr=3e-4, betas=(.9, .999),
            eps=1e-7, amsgrad=True)

        for epoch in range(num_epochs):
            # train
            # epoch_loss = 0.
            for X_train, y_train in dgen:
                optimizer.zero_grad()
                y_pred = model(X_train)
                loss = loss_fn(y_pred, y_train)
                loss.backward()
                optimizer.step()
                # epoch_loss += loss.item()
            # early stopping
            if y_valid is not None:
                with torch.no_grad():
                    tmp = loss_fn(model(X_valid), y_valid).item()
                    if tmp >= valid_loss:
                        wait += 1
                    else:
                        valid_loss = tmp
                        wait = 0
                    if wait > patience:
                        break

        # load datasets
        dgen_test = torch.utils.data.DataLoader(
            ds_test, batch_size=len(ds_test), shuffle=False)
        for X_test, y_test in dgen_test:
            break

        dgen_train = torch.utils.data.DataLoader(
            ds_train, batch_size=len(ds_train), shuffle=False)
        for X_train, y_train in dgen_train:
            break

        # compute metrics
        y_pred = torch.argmax(model(X_test), dim=1)
        res_test = {
            "num": len(y_pred),
            "acc": sklearn.metrics.accuracy_score(y_test, y_pred),
            "acc-balanced": sklearn.metrics.balanced_accuracy_score(
                y_test, y_pred),
            "f1": sklearn.metrics.f1_score(y_test, y_pred, average='micro'),
            "f1-balanced": sklearn.metrics.f1_score(
                y_test, y_pred, average='macro'),
        }
        y_pred = torch.argmax(model(X_train), dim=1)
        res_train = {
            "num": len(y_pred),
            "acc": sklearn.metrics.accuracy_score(y_train, y_pred),
            "acc-balanced": sklearn.metrics.balanced_accuracy_score(
                y_train, y_pred),
            "f1": sklearn.metrics.f1_score(y_train, y_pred, average='micro'),
            "f1-balanced": sklearn.metrics.f1_score(
                y_train, y_pred, average='macro'),
        }
        # save results
        results.append({
            "task": downstream_task,
            "epochs": epoch + 1,
            "test": res_test,
            "train": res_train
        })

    # done
    return results
